python_modules/libraries/dagster-components/dagster_components/core/component.py

```diff
@@ -193,12 +196,91 @@ def get_component_types_in_module(
 T = TypeVar("T")
 
 
+@contextlib.contextmanager
+def _add_to_python_path(path: Path) -> Iterator[None]:
+    sys.path.append(path.as_posix())
+    try:
+        yield
+    finally:
+        sys.path.pop()
+
+
+@dataclass
+class DefinitionsModuleCache:
+    """Context for loading an entire component hierarchy in a code location.
+    Stores resources and a cache to ensure we don't load the same component multiple times.
+    """
+
+    resources: Mapping[str, object]
+
+    def build_defs_from_component_path(self, path: Path) -> Definitions:
+        """Loads a set of Dagster definitions from a components Python module.
+        Eventually would like to move off this entirely to use build_defs_from_component_module.
+
+        Args:
+            path (Path): The path to the components Python module.
+
+        Returns:
+            Definitions: The set of Dagster definitions loaded from the module.
+        """
+        # We need to make sure the project root is in the Python path, this isn't always the case
+        # when loading components from a path.
+        with _add_to_python_path(path.parent.parent.parent):
```

Comment:
Not super proud of this, but don't want to alter the upstream behavior which solely imports components by path (e.g. doesn't make any determinations on whether they are on the `PATH` as importable modules or not). Can maybe change in a stacked PR?

Reply:
Can we change that _before_ merging this? This is pretty rough